src/downloader/sentry_client_reactor.rs

```diff
@@ -191,64 +191,144 @@ impl Drop for SentryClientReactor {
     }
 }
 
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+enum EventLoopStreamId {
+    Send,
+    Receive,
+    Stop,
+}
+
+#[derive(Debug)]
+enum EventLoopStreamResult {
+    Send(u32),
+    Receive(MessageFromPeer),
+    Stop(()),
+}
+
+type EventLoopStream = Pin<Box<dyn Stream<Item = anyhow::Result<EventLoopStreamResult>> + Send>>;
+
+// dropping this causes dropping the senders and triggers an end of stream event on the receivers end
+struct EventLoopReceiveMessagesSendersDropper {
+    receive_messages_senders: Arc<RwLock<HashMap<EthMessageId, broadcast::Sender<Message>>>>,
+}
+
+impl Drop for EventLoopReceiveMessagesSendersDropper {
+    fn drop(&mut self) {
+        // drop shared senders so that existing receive_messages() streams end
+        self.receive_messages_senders.write().clear();
+    }
+}
+
 impl SentryClientReactorEventLoop {
     async fn run(&mut self) -> anyhow::Result<()> {
+        let mut sentry = self.sentry.take().unwrap();
```

Comment:
This will panic on the second `run` - `run` should consume `SentryClientReactorEventLoop`.